src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -52,80 +74,170 @@ private static SnowflakeDateTimeFormat createDateTimeFormatter() {
+    if (input instanceof String) {
+      try {
+        return objectMapper.readTree((String) input);
+      } catch (JsonProcessingException e) {
+        throw valueFormatNotAllowedException(input, snowflakeType, "Not a valid JSON");
+      }
+    } else if (isAllowedVariantType(input)) {
+      return objectMapper.valueToTree(input);
     }
 
-    if (output.length() > BYTES_16_MB) {
-      throw new SFException(
-          ErrorCode.INVALID_ROW,
-          input.toString(),
-          String.format("Variant too long: length=%d maxLength=%d", output.length(), BYTES_16_MB));
+    throw typeNotAllowedException(
+        input.getClass(),
+        snowflakeType,
+        new String[] {
+          "String",
+          "Primitive data types and their arrays",
+          "java.time.*",
+          "List<T>",
+          "Map<String, T>",
+          "T[]"
+        });
+  }
+
+  /**
+   * Validates and parses input as JSON. All types in the object tree must be valid variant types,
+   * see {@link DataValidationUtil#isAllowedVariantType}.
+   *
+   * @param input Object to validate
+   * @return JSON string representing the input
+   */
+  static String validateAndParseVariant(Object input) {
+    String output = validateAndParseVariantAsJsonTree(input, "VARIANT").toString();
+    int stringLength = output.getBytes(StandardCharsets.UTF_8).length;
+    if (stringLength > MAX_VARIANT_LENGTH) {
+      throw valueFormatNotAllowedException(
+          input,
+          "VARIANT",
+          String.format(
+              "Variant too long: length=%d maxLength=%d", stringLength, MAX_VARIANT_LENGTH));
     }
     return output;
   }
 
   /**
-   * Expects an Array or List object
+   * Validates that passed object is allowed variant type. For non-trivial types like Maps, arrays
+   * or Lists, it recursively traverses the object tree and validates that all types in the tree are
+   * also allowed. Allowed Java types:
    *
-   * @param input the input data, must be able to convert to String
+   * <ul>
+   *   <li>primitive types (int, long, boolean, ...)
+   *   <li>String
+   *   <li>BigInteger
+   *   <li>BigDecimal
+   *   <li>LocalTime
+   *   <li>OffsetTime
+   *   <li>LocalDate
+   *   <li>LocalDateTime
+   *   <li>OffsetDateTime
+   *   <li>ZonedDateTime
+   *   <li>Map<String, T> where T is an allowed variant type
+   *   <li>List<T> where T is an allowed variant type
+   *   <li>primitive arrays (char[], int[], ...)
+   *   <li>T[] where T is an allowed variant type
+   * </ul>
+   *
+   * @param o Object to validate
+   * @return If the passed object is allowed for ingestion into a VARIANT column
    */
-  static String validateAndParseArray(Object input) {
-    if (!input.getClass().isArray() && !(input instanceof List)) {
-      throw new SFException(
-          ErrorCode.INVALID_ROW, input, "Input column must be an Array or List object.");
+  static boolean isAllowedVariantType(Object o) {
```

Comment:
These checks are great but it might be more readable to have brackets for each of the conditions. I'm nervous that someone adding a check or changing this might inadvertently goof up a check otherwise.

Reply:
Braces added

Reply:
It still looks a bit messy, could you add comment for each `if` statement?

Reply:
And this is used not only for VARIANT, should we call it `isAllowedSemiStructuredType`?

Reply:
Method renamed, I also restructured the if/else conditions to make the logic easier to follow + added comments.